WIP: Use Typescript compiler API's 'transpileModule'

Thorn1089 opened the pull request:
The custom `transpileModule` defined in this plugin creates its own `CompilerHost` that can only resolve a single file -- the file being transpiled. This means that when the Typescript infrastructure wants to look up definition files for a particular target, they aren't resolved. A concrete example is when trying to target ES6/ES2015 to use native Promises and async/await. `tsc` on the command-line will succeed, but `brunch build` will fail as the .d.ts files with the ES2015 features will not be resolved.

The ignored errors from `transpile.js` have been moved to `index.js` and the same filtering of diagnostics applied. Source maps are still respected based on configuration. ~~I also pulled in the changes from the outstanding PR to move Typescript itself to a peer dependency (requiring clients to declare the dependency as a `devDependency` themselves).~~ Backed out this change. Typescript after 2.2.2 has an issue where `node_modules` dependencies don't get compiled by Brunch anymore. I'll try and tackle that separately, as I do think moving to a `peerDependency` is the right approach, but the plugin won't function in most cases currently when doing so.

oeb25 committed:
Make TypeScript a peerDependency instead of a static dependency to allow user defined TypeScirpt versions

TomRK1089 commented:
I just marked this as WIP -- it looked good locally on simple test projects, but I'm running into issues when trying to use the modified version and compile required modules from `node_modules`. Unclear if it's the changes or how `yarn link` was used to test with the modified plugin.

Thorn1089 committed:
Uses built-in transpilation API from Typescript compiler

Rather than reimplementing the transpileModule function, use it directly. Trying to create a limited CompilerHost to feed to the tooling means that specifying alternate targets besides ES5 does not function correctly. For example, prior to this fix, even if tsconfig.json specified "target": "es2015" to include a native Promise implementation, this plugin would error out on async/await constructs because the relevant lib.d.ts files could not be resolved by the CompilerHost.

Thorn1089 committed:
Bumps Typescript dependency to 2.2.2, latest that works with Brunch

2.3.0 and later cause an issue where dependencies from node_modules are no longer built by Brunch. The cause is not known and should be handled in a separate PR.

TomRK1089 commented:
Issue with `node_modules` diagnosed -- newer versions of Typescript cause this issue. 2.2.2 is the newest safe version, which is still shockingly old given a 3.0 RC was just announced, but that's a problem for a different PR.

Thorn1089 committed:
Removes targetExtension property that breaks dependency compilation when used with Brunch v2.10.x

Thorn1089 committed:
Merge branch 'master' of https://github.com/oeb25/typescript-brunch

# Conflicts:
#	package.json

paulmedynski commented:
I have typescript-brunch working with TypeScript 3.1.0.  I patched `transpile.js` to use `ts.createCompilerHost()` to get the default host, and then overwrite the `getSourceFile()` and `writeFile()` functions.  Everything else stays default, and compiling to ES2017 works fine for me.  I nuked the hardcoded `noLib` as well, since it isn't necessary when `getSourceFile()` allows reading of other files.  I also patched `index.js` to report error line/col numbers correctly.

I took a look at `ts.traspileModule()` in TS 3.1.0, and it appears to have been based on this project's `transpileModule()`, complete with hardcoded `noLib` and clearing of the `lib` option, so I don't think that implementation will work for Brunch.  Perhaps we don't want to use `ts.transpileModule()` after all.
